### PR TITLE
V2: Modify DataLoader to drop last batch if its size is 1, to avoid batch norm issues

### PR DESCRIPTION
**Pull request by jonwzheng**
## Description
Calling `nn.BatchNorm1d` on a batch of size 1 will lead to a ValueError. This can occur if the remainder of the dataset and the batch size is 1 (for example, dataset has 101 samples and the batch size is 10). 

## Example / Current workflow
From a clean install of Chemprop, install the FreeSolv dataset and run the command:
`chemprop train --data-path data/freesolv.csv --task-type regression --output-dir freesolv`
which will lead to a ValueError. Traceback shows that this is due to calling `batch_norm` for a batch of size 1.

## Bugfix / Desired workflow
This PR implements one fix suggested in https://github.com/pytorch/pytorch/issues/4534, which is to drop the last batch if its size is 1. This also includes a warning so that the user can see that this happened.

## Questions
do we expect any problems with using `len(dataset)` for `MulticomponentDataset` and `ReactionDataset`?

this is an overly-aggressive option; what if `batch_norm=False`? is there a smooth way to only activate this if batch norm is desired?

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

**Review by KnathanM — APPROVED** [review]
Thank you Jonathan for adding this fix.

Potentially there could be an issue with dropping a datapoint and using `len(dataset)` but I don't think we need to worry because we are the only ones who use `drop_last`. And when training, our code never sees the whole dataset, just the self sufficient batches handed to it by the dataloader. Additionally, best I can tell, `trainer.predict(model, dataloader)` ignores the batchsize of the dataloader and does predictions one at a time so no datapoints are left out. 

I couldn't think of a smooth way to only drop_last if there is batch norm because of the separation between models and dataloaders. The dataloader never sees the model and so can't know if the model has batch normalization. It would be up to the model to respond if the dataloader gives it a batch size of 1, but I don't want to add an `if len(batch) == 1` to the model because it could slow down training. 

I think the case where a user doesn't want batch normalization, has a dataset and batch size that % == 1, and cares about training on that 1 datapoint be rare and can be ignored.

As a side note for myself, I wondered about using `logging` instead of `warnings` because that is what the CLI uses, but then I remembered the dataloaders are also used in notebooks that don't have `logging`. 

Finally, I ran your changes through the black formatting.

**Comment by jonwzheng** [comment]
Now that `batch_norm` is off by default, is this something that we want to revisit?
